projects/core/components/primitive-calendar/primitive-calendar.component.ts

```diff
@@ -104,12 +101,22 @@ export class TuiPrimitiveCalendarComponent {
     }
 
     getItemRange(item: TuiDay): TuiRangeState | null {
-        const {value, hoveredItem} = this;
+        const {hoveredItem} = this;
+        // Fix problem with Array.isArray
+        // See: https://github.com/microsoft/TypeScript/issues/17002
+        const value = this.value as Exclude<
+            TuiDay | TuiDay[] | TuiDayRange | readonly TuiDay[] | null,
+            readonly TuiDay[]
+        >;
 
         if (!value) {
             return null;
         }
 
+        if (Array.isArray(value)) {
```

Comment:
Move this after the `TuiDay` check and change to:
```ts
        if (!(value instanceof TuiDayRange)) {
            return value.find(day => day.daySame(item)) ? TuiRangeState.Single : null;
        }
```

Reply:
fixed